develop/blockchains/maintenance/storage-migrations.md

```diff
@@ -1,4 +1,183 @@
+
+- **[`post_upgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.Hooks.html#method.post_upgrade){target=\_blank}** - after the migration completes, `post_upgrade` validates its success. It receives the state data captured by `pre_upgrade` and uses this to verify that the migration executed correctly. This function checks for storage consistency and ensures all data transformations occurred as expected. Like `pre_upgrade`, it operates exclusively in testing environments and should not modify storage
+
+### Migration Logic
+
+When implementing the actual migration logic, your code needs to handle several key responsibilities. Your migration implementation should:
+
+- Read existing storage values in their original format
+- Transform data to match the new format
+- Write updated values back to storage
+- Calculate and return consumed weight
+
+### Migration Structure
+
+Storage migrations can be implemented using two approaches. The first method involves directly implementing `OnRuntimeUpgrade` on structs. This approach requires manually checking the on-chain storage version against the new [`StorageVersion`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/struct.StorageVersion.html){target=\_blank} and executing the transformation logic only when the check passes. This version verification prevents multiple executions of the migration during subsequent runtime upgrades.
+
+The recommended approach is to implement [`UncheckedOnRuntimeUpgrade`](https://paritytech.github.io/polkadot-sdk/master/frame_support/traits/trait.UncheckedOnRuntimeUpgrade.html){target=\_blank} and wrap it with [`VersionedMigration`](https://paritytech.github.io/polkadot-sdk/master/frame_support/migrations/struct.VersionedMigration.html){target=\_blank}. `VersionedMigration` implements `OnRuntimeUpgrade` and handles storage version management automatically, following best practices and reducing potential errors.
+
+`VersionedMigration` requires five type parameters:
+
+- `From` - the source version for the upgrade
+- `To` - the target version for the upgrade
+- `Inner` - the `UncheckedOnRuntimeUpgrade` implementation
+- `Pallet` - the pallet being upgraded
+- `Weight` - the runtime's [`RuntimeDbWeight`](https://paritytech.github.io/polkadot-sdk/master/frame_support/weights/struct.RuntimeDbWeight.html){target=\_blank} implementation
+
+```rust
+--8<-- 'https://raw.githubusercontent.com/paritytech/polkadot-sdk/refs/tags/polkadot-stable2409-1/substrate/frame/examples/single-block-migrations/src/migrations/v1.rs:0:122'
```

Comment:
I know this might be a pain in the ass, but I think we should grab this remote file, taking only the lines we consider important. For example, all the comments at the very beginning are just making visual noise for somebody who reads this code snippet.

Reply:
And also, somewhere a tiny paragraph should be added to explain what's exactly this coding is doing